cli/internal/core/engine.go

```diff
@@ -257,3 +268,77 @@ func (e *Engine) AddDep(fromTaskID string, toTaskID string) error {
 
 	return nil
 }
+
+// ValidatePersistentDependencies checks if any task dependsOn persistent tasks and throws
+// an error if that task is actually implemented
+func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) error {
+	var validationError error
+
+	// Adding in a lock because otherwise walking the graph can introduce a data race
+	// (reproducible with `go test -race`)
+	var sema = util.NewSemaphore(1)
+
+	errs := e.TaskGraph.Walk(func(v dag.Vertex) error {
+		vertexName := dag.VertexName(v) // vertexName is a taskID
+
+		// No need to check the root node if that's where we are.
+		if strings.Contains(vertexName, ROOT_NODE_NAME) {
+			return nil
+		}
+
+		// Aquire a lock, because otherwise walking this group can cause a race condition
+		// writing to the same validationError var defined outside the Walk(). This shows
+		// up when running tests with the `-race` flag.
+		sema.Acquire()
+		defer sema.Release()
+
+		currentPackageName, currentTaskName := util.GetPackageTaskFromId(vertexName)
+
+		// For each "downEdge" (i.e. each task that _this_ task dependsOn)
+		// check if the downEdge is a Persistent task, and if it actually has the script implemented
+		// in that package's package.json
+		for dep := range e.TaskGraph.DownEdges(vertexName) {
+			depTaskID := dep.(string)
+			// No need to check the root node
+			if strings.Contains(depTaskID, ROOT_NODE_NAME) {
+				return nil
+			}
```

Comment:
The "ROOT_NODE" in the graph is different from a task that comes from the root workspace.

I'm not sure why we add this root node, but it's the `___ROOT___` in the graph:

![graph](https://user-images.githubusercontent.com/490968/200943101-fc6abf25-11cc-4b52-8335-4ea09b0f0a58.png). 

I've added a test case for the example you gave and can validate that it fails if I don't do this early return. The failure is incidental (getting the packageName and taskName from `"___ROOT___"` fails, because it's not in the form of `xx#yy`, but I did some some time trying to see if that was a valid thing to try to address, and I think it isn't. We also do the same early exit when walking the graph for execution:

https://github.com/vercel/turbo/blob/a8b6551bbcfdc3bb3e30f727908b857210e36d79/cli/internal/core/engine.go#L87-L90